HBASE-23371 [HBCK2] Provide client side method for removing ghost regions in meta.

wchevreuil opened the pull request:
Figured there were many similarities with the already existing _addFsRegionsMissingInMeta_ and _reportMissingRegionsInMeta_, so had worked on some refactoring in order to avoid code duplication.

saintstack commented on hbase-hbck2/README.md:
```diff
@@ -174,13 +174,46 @@ Command:
+   directories on the file system. Needs hbase:meta
+   to be online. For each table name passed as parameter, performs diff
+   between regions available in hbase:meta and region dirs on the given
+   file system, removing extra regions from meta with no matching directory.
```
Removing regions from meta? Favoring the filesystem view?

wchevreuil replied:
Yeah (see Daisuke's comments next). This is similar to the previous _addFsRegionsMissingInMeta_ command, where we assume the filesystem to be the source of truth. Given the "ghost" region in meta is useless without a related dir in the filesystem, here deletion would have no risk of dataloss.

saintstack commented on hbase-hbck2/README.md:
```diff
@@ -174,13 +174,46 @@ Command:
    '--fix' option. Pass a table name to check for replication barrier and
    purge if '--fix'.
 
+ reportExtraRegionsInMeta <NAMESPACE|NAMESPACE:TABLENAME>...
```
You don't want to add one method and then add a --fix flag for case of where you want the command to actually perform fixes? (filesystem does this -- reporting if no flags and then fixing if a --fix passed).

saintstack replied:
This cuts down some on the number of commands operator sees.

wchevreuil replied:
Yep, makes sense, let me refactor it accordingly. Coming on next commit.

wchevreuil replied:
Yep, makes sense, let me refactor it accordingly. Coming on next commit.

daisukebe commented:
Hi @wchevreuil, couple of comments based on my test. 

Unless the region is overlapped with other regions in meta, uses should use `assign '<region>'` command as a viable action to take because it generates a new empty region directory appropriately via `org.apache.hadoop.hbase.regionserver.HRegionFileSystem#writeRegionInfoFileContent`. 

Otherwise, in a case where the region is overlapped with others in meta, we should remove the region's entry from meta as assuming it does not have valid data in HDFS.

Thoughts?

daisukebe commented on hbase-hbck2/src/main/java/org/apache/hbase/HBCK2.java:
```diff
@@ -99,8 +97,10 @@
   private static final String REPORT_MISSING_REGIONS_IN_META = "reportMissingRegionsInMeta";
+  static final String REMOVE_EXTRA_REGIONS_IN_META_FROM_TABLES =
+    "removeExtraRegionsFromMeta";
+  static final String REPORT_EXTRA_REGIONS_IN_META = "reportExtraRegionsInMeta";
```
People in the real world, including me, are still using hbck1 as a trusted report to deal with inconsistencies. Good to have this functionality in HBCK2 too!

wchevreuil replied:
> Unless the region is overlapped with other regions in meta, uses should use assign '<region>' command as a viable action to take because it generates a new empty region directory appropriately via org.apache.hadoop.hbase.regionserver.HRegionFileSystem#writeRegionInfoFileContent.

Thanks for pointing this out, @daisukebe ! Let me add this valuable insight to the command instructions and "recipes" section of the README.

> People in the real world, including me, are still using hbck1 as a trusted report to deal with inconsistencies

Yeah, hbck2 is not targeted to be a reporting tool, but as part of work for both this new command and the _addFsRegionsMissingInMeta_ is to figure out which regions these need to fix, had then decided to expose the _checking_ logic as a reporting command that can be executed on its own.

wchevreuil commented:
New commit pushed to this PR. Changes:
1) Made a single command "extraRegionsInMeta" with a "-f/-fix" option as suggested by Stack;
2) Added remarks about "assigns" command in the help text, as pointed out by Daisuke;

daisukebe commented on hbase-hbck2/src/main/java/org/apache/hbase/HBCK2.java:
```diff
@@ -374,6 +359,8 @@ private static String getCommandUsage() {
     writer.println();
     usageBypass(writer);
     writer.println();
+    usagetExtraRegionsInMeta(writer);
```
Typo: `usageExtraRegionsInMeta`?

wchevreuil replied:
Will be fixed on next commit.

daisukebe commented on hbase-hbck2/src/main/java/org/apache/hbase/HBCK2.java:
```diff
@@ -490,6 +477,33 @@ private static void usageReplication(PrintWriter writer) {
     writer.println("   purge if '--fix'.");
   }
 
+  private static void usagetExtraRegionsInMeta(PrintWriter writer) {
```
Same typo: `usageExtraRegionsInMeta`

wchevreuil replied:
Will be fixed on next commit.

joshelser commented on hbase-hbck2/src/main/java/org/apache/hbase/FsRegionsMetaRecoverer.java:
```diff
@@ -33,19 +38,21 @@
 /**
- * This class implements the inner works required for check and recover regions that wrongly
- * went missing in META.
+ * This class implements the inner works required for checking and recover regions that wrongly
```
nit: recovering (to match "checking")

wchevreuil replied:
Coming on next commit.

joshelser commented on hbase-hbck2/src/main/java/org/apache/hbase/FsRegionsMetaRecoverer.java:
```diff
@@ -70,61 +77,207 @@ public FsRegionsMetaRecoverer(Configuration configuration) throws IOException {
+
+  List<RegionInfo> findExtraRegionsInMETA(String table) throws IOException {
+    InternalMetaChecker<RegionInfo> extraChecker = new InternalMetaChecker<>();
+    return extraChecker.checkRegionsInMETA(table, (regions,dirs) -> {
```
Ultimately this is the same logic as `findMissingRegionsInMETA`.

For types `A`, `B`, and `C`: take a `Collection<A>`, `Collection<B>`, a function `f(A) -> C` and `f(B) -> C`. We want to find all elements in `A` that don't exist in `B`. In Set theory, this is the difference: https://en.wikipedia.org/wiki/Complement_(set_theory)#Relative_complement

For this code, we have `Path` (A) and `RegionInfo` (B) and we want to transform them into a common type `String` (C) which is the encoded region name.

Pseudocode-y
`setComplement(Set<RegionInfo> regionsInMeta, Set<Path> regionsInHdfs, Function<RegionInfo,String> regionInfoToName, Function<Path, String> pathToRegionName)`
`setComplement(Set<Path> regionsInHdfs, Set<RegionInfo> regionsInMeta, Function<Path,String> ..., Function<RegionInfo,String> ..)`

If it's not terribly difficult, could the underlying logic be consolidated?

wchevreuil replied:
Yeah, I guess we could basically move the whole loop to InternalMetaChecker. Let me work on that and try get it improved in the next commit.

joshelser replied:
This is nice!

joshelser commented on hbase-hbck2/src/main/java/org/apache/hbase/HBCK2.java:
```diff
@@ -191,76 +189,63 @@ int setRegionState(ClusterConnection connection, String region,
+          result.put(e.getKey(),
+            e.getValue().stream().map(r->r.getEncodedName()).collect(Collectors.toList()));
+          if(fix && e.getValue().size()>0){
+            toFix.add(e.getKey().getNameWithNamespaceInclAsString());
```
Why convert `TableName` to a `String`? It seems like when you end up using this String, it would be easier (less error-prone) to have the `TableName`.

joshelser replied:
Answered elsewhere -- string is used to send both table name or namespace

joshelser commented on hbase-hbck2/src/main/java/org/apache/hbase/FsRegionsMetaRecoverer.java:
```diff
@@ -70,61 +77,207 @@ public FsRegionsMetaRecoverer(Configuration configuration) throws IOException {
+            if(namespacesOrTables==null || namespacesOrTables.isEmpty()){
+              return true;
+            } else {
+              Optional<String> findings = namespacesOrTables.stream().filter(
```
This is the other half: if we have a `TableName` here, can we simplify this check?

wchevreuil replied:
Problem is that I'm allowing _namespacesOrTables_ value to be a namespace, in which case, all tables under the given namespace should be checked. If we make it a _TableName_ type, I guess we won't be able to pass a namespace as value.

joshelser commented on hbase-hbck2/src/main/java/org/apache/hbase/FsRegionsMetaRecoverer.java:
```diff
@@ -70,61 +77,207 @@ public FsRegionsMetaRecoverer(Configuration configuration) throws IOException {
+        }
+      } catch (IOException | InterruptedException e) {
+        LOG.error("ERROR executing thread: ", e);
+        throw new IOException(e);
```
Unnecessary wrap with another `IOE`.

wchevreuil replied:
Coming on next commit.

joshelser commented on hbase-hbck2/src/main/java/org/apache/hbase/FsRegionsMetaRecoverer.java:
```diff
@@ -70,61 +77,207 @@ public FsRegionsMetaRecoverer(Configuration configuration) throws IOException {
+
+  public static void main(String[] args){
+    System.out.println("hi");
+  }
```
Remove :)

wchevreuil replied:
Oops.

wchevreuil commented:
Ping @joshelser @saintstack @daisukebe , any further thoughts on this one?

joshelser commented on hbase-hbck2/src/main/java/org/apache/hbase/FsRegionsMetaRecoverer.java:
```diff
@@ -70,61 +78,203 @@ public FsRegionsMetaRecoverer(Configuration configuration) throws IOException {
+                tableName.getNameWithNamespaceInclAsString());
+            }
+          }
+          for(Future<List<String>> f : futures){
```
Instead of calling `future.get()` here, what about `future.isDone()` and sending the `Future<List<String>>` back to the caller? It would simplify the logic here and still let you shut down the `ExecutorService`.

Your method return type could just be `Map<String,Future<List<String>>>`, mapping "namespaceOrTable" to a `Future` with the exception or result of computation.

wchevreuil replied:
Maybe I'm not following your thought, but I think doing `future.isDone()` would require extra logic to keep iterating through all futures that are not yet completed, or introduce a latch for control. I think it would still be simpler to keep using `future.get()`, but together with your suggestion for returning `Map<String,Future<List<String>>>`. Let me give it a try.

joshelser replied:
Yeah, `f.get()` is going to be a blocking call, so you would want to have some kind of back-off/sleep/recheck loop here instead (if you choose to take my suggestion ;)).

Something like:
```
ArrayList<Future<List<String>>> completedFutures = new ArrayList<>();
Future<List<String>> current = null;
while (!futures.isEmpty()) {
  Iterator<Future<List<String>>> iter = futures.iterator();
  while (iter.hasNext()) {
    current = iter.next();
    if (current.isDone()) {
      iter.remove();
      completedFutures.add(current);
    }
  }
  if (!futures.isEmpty()) {
    Thread.sleep(2000);
  }
}
```

wchevreuil replied:
Ok, changed to return a `List<Future<List<String>>>`. I realised `future.get()` wouldn't make sense here, I would still need to call it in the caller to actually get results and potential errors.

wchevreuil replied:
Sorry @joshelser , had missed your last comment while I was working on my last change. Let me know what you think on this last one.

joshelser replied:
No worries! My suggestion would retain the semantics that you have already (where this method is blocking on completion). You could also just make this return fast, leaving  the author to wait on f.get(). But yeah, leave it up to the caller to decide what is done with errors, IMO (e.g. maybe we find later that HBCK2 should auto-retry a number of times?)

I forget the semantics of `ExecutorService.shutdown()` -- I think that might just reject any new submissions and shut down once everything currently running is done (not creating a "how do I make sure the ES is closed" problem for you).

joshelser commented on hbase-hbck2/src/main/java/org/apache/hbase/FsRegionsMetaRecoverer.java:
```diff
@@ -70,61 +78,203 @@ public FsRegionsMetaRecoverer(Configuration configuration) throws IOException {
+  public class ListUtils<T1, T2> {
+    public List<T1> complement(List<T1> list1, List<T2> list2,
+        Function<T2, String> convertFromBase,
+        Function<T1, String> convertFromComparing) throws IOException {
```
Weird to have the order of T1 and T2 reversed (1 & 4, then 2 & 3). What about `List<T1>, List<T2>, Function<T1,String>, Function<T2,String>`?

joshelser commented on hbase-hbck2/src/main/java/org/apache/hbase/FsRegionsMetaRecoverer.java:
```diff
@@ -70,61 +78,203 @@ public FsRegionsMetaRecoverer(Configuration configuration) throws IOException {
+        Function<T1, String> convertFromComparing) throws IOException {
+      final List<T1> extraRegions = new ArrayList<>();
+      HashSet<String> baseSet = list2.stream().map(info ->
+        convertFromBase.apply(info)).collect(Collectors.toCollection(HashSet::new));
```
`convertFromBase` and `convertFromComparing` isn't clear to me what they're doing. Maybe just name them according to the type, e.g. `convertT1` and `convertT2`, since the types are opaque anyways?

wchevreuil replied:
Indeed confusing, even I had troubled to figure it out after not touching this code for a month. Changing this and the params order mentioned on above commented on next commit.

joshelser commented on hbase-hbck2/src/main/java/org/apache/hbase/HBCK2.java:
```diff
@@ -191,76 +189,63 @@ int setRegionState(ClusterConnection connection, String region,
+      result = fsRegionsMetaRecoverer.addMissingRegionsInMetaForTables(
+        formatNameSpaceTableParam(nameSpaceOrTable));
+    } catch (IOException e) {
+      LOG.error("Error adding missing regions: ", e);
```
Include `nameSpaceOrTable` in the message?

wchevreuil replied:
We don't have how to know for which of the passed namespace or table we are really erroring here, so ain't sure adding the list of all passed namespaces/tables would add much value here.

joshelser commented on hbase-hbck2/src/main/java/org/apache/hbase/HBCK2.java:
```diff
@@ -843,22 +868,34 @@ private static String toString(List<?> things) {
   private String formatMissingRegionsInMetaReport(Map<TableName,List<Path>> report) {
+    Function<Path,String> resolver = r -> r.getName();
+    String message = "Missing Regions for each table:\n\t";
+    return formatReportMessage(message, new HashMap<>(report), resolver);
```
Just cast this instead of creating a brand new HashMap?

joshelser commented on hbase-hbck2/src/main/java/org/apache/hbase/hbck1/HBaseFsck.java:
```diff
@@ -288,7 +288,7 @@
    * Options
    ***********/
-  private static boolean details = false; // do we display the full report
+  private static boolean details = false; // do we display the full execute
```
Unintentional change of `s/report/execute/` here and below?

wchevreuil replied:
Definitely yeah. Reverting this.

wchevreuil commented on hbase-hbck2/README.md:
```diff
@@ -148,6 +148,29 @@ Command:
+   regions dirs in the filesystem, if not existing.
+   An example triggering extra regions report for tables 'table_1'
+   and 'table_2', under default namespace:
+     $ HBCK2 extraRegionsInMeta default:table_1 default:table_2
```
Yeah, in the same way as done for the already existing _addFsRegionsMissingInMeta_ and _reportMissingRegionsInMeta_ methods. Since we allow for passing namespaces as well, if we allow for simply passing the table name for tables in default namespace, we would need to do an extra verification for the existence of such table in default namespace, and if no such table found, then fallback to consider it as namespace. And then there's also possibility for a table name in default namespace have same value as a given namespace, which would then prohibit this namespace to be passed. All that made me thought it was better to require explicitly passing the default namespace prefix when operating over specific table.

joshelser replied:
Ok! Just double checking.

wchevreuil commented:
Addressed your suggestions on last commit, @joshelser , am trying to figure out how to make this PR applicable to master now, given all the conflicts. Tried a rebase, but it didn't work, apparently.

wchevreuil commented:
Looks like my last rebase worked. Waiting on the build results.

wchevreuil commented:
Ok, only remaining checkstyle issue complains about length of HBCK2.doCommandLine, which unless we do a full re-design of HBCK2 to implement the Command design pattern, I don't think it's addressable, as we will be constantly adding more options there.

wchevreuil commented:
Signed-off-by: Josh Elser <elserj@apache.org>

joshelser commented:
🎉
